src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -243,47 +243,35 @@ private bool StartCore(ProcessStartInfo startInfo)
                 }
             }
 
-            int childPid = -1, stdinFd = -1, stdoutFd = -1, stderrFd = -1, result = -1;
+            int childPid, stdinFd, stdoutFd, stderrFd, result = -1;
             string[] envp = CreateEnvp(startInfo);
             string cwd = !string.IsNullOrWhiteSpace(startInfo.WorkingDirectory) ? startInfo.WorkingDirectory : null;
 
-            filename = ResolvePath(startInfo.FileName); 
-            if (!string.IsNullOrEmpty(filename))
-            {
-                argv = ParseArgv(startInfo);
+            filename = ResolvePath(startInfo.FileName);
+            argv = ParseArgv(startInfo);
 
-                // Invoke the shim fork/execve routine.  It will create pipes for all requested
-                // redirects, fork a child process, map the pipe ends onto the appropriate stdin/stdout/stderr
-                // descriptors, and execve to execute the requested process.  The shim implementation
-                // is used to fork/execve as executing managed code in a forked process is not safe (only
-                // the calling thread will transfer, thread IDs aren't stable across the fork, etc.)
-                result = Interop.Sys.ForkAndExecProcess(
+            // Invoke the shim fork/execve routine.  It will create pipes for all requested
+            // redirects, fork a child process, map the pipe ends onto the appropriate stdin/stdout/stderr
+            // descriptors, and execve to execute the requested process.  The shim implementation
+            // is used to fork/execve as executing managed code in a forked process is not safe (only
+            // the calling thread will transfer, thread IDs aren't stable across the fork, etc.)
+            result = Interop.Sys.ForkAndExecProcess(
                     filename, argv, envp, cwd,
                     startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
                     out childPid,
                     out stdinFd, out stdoutFd, out stderrFd, shouldThrow: !startInfo.UseShellExecute);
-            }
 
-            if (result != 0)
+            if (result != 0 && startInfo.UseShellExecute)
```

Comment:
This logic is a little confusing.  Am I right that if result != 0 then UseShellExecute must be true (as otherwise an exception would have already been thrown)?  If so, seems like we should assert that and then remove the check from the if.

Reply:
Once the build is completed, I will push another commit that aims to simplify the logic.

I will push the commit in about 10mins